src/Utils/obiDialogParser.ts

```diff
@@ -112,51 +123,141 @@ export class ObiDialogParser {
         return luMap
     }
 
-    private async getTrainDialogsfromOBIDialog(obiDialog: OBITypes.OBIDialog): Promise<CLM.TrainDialog[]> {
-
-        let trainDialogs: CLM.TrainDialog[] = []
+    /**
+     * Walks the dialog tree from the given node.  Validates types of nodes in the tree and returns an
+     * in-memory representation of the tree.
+     * Tree construction is slightly complicated since child nodes can be referenced in `rules` or `steps`.
+     */
+    private async collectDialogNodes(obiDialog: OBITypes.OBIDialog): Promise<ObiDialogNode> {
+        let node: ObiDialogNode = new ObiDialogNode(obiDialog)
         if (obiDialog.rules) {
-            for (const rule of obiDialog.rules) {
-                if (rule.$type === OBIRuleType.INTENT_RULE) {
+            await this.collectDialogRuleChildren(node, obiDialog.rules)
+        }
+        if (obiDialog.steps) {
+            await this.collectDialogStepChildren(node, obiDialog.steps)
+        }
+        return node
+    }
 
-                    const textVariations = this.getTextVariations(rule.intent!)
-                    const extractorStep: CLM.TrainExtractorStep = {
-                        textVariations: textVariations
-                    }
-                    if (!rule.steps) {
+    // Collects dialog tree nodes from `Microsoft.IntentRule` elements in the dialog `rules` section.
+    private async collectDialogRuleChildren(node: ObiDialogNode, rules: OBITypes.MicrosoftIRule[]) {
```

Comment:
Don't have much context on the OBI stuff but it's weird for recursive functions (`collectDialogRuleChildren`, `collectDialogStepChildren`) to modify their inputs `node.children.push(child)`

I would expect given a node and rules to return the children, and then next statement can choose to add the children to the node. Similar to other conversation about trying for pure functions, although I assume this codebase past that point so maybe it's too much to change.

Reply:
Maybe this is more of a Java / C++ pattern, but in my experience it's pretty common to add things to collections via out-parameters in recursive functions that have this kind of structure.  Out-params are more obvious in C++ where you'd have a pointer parameter.

Happy to change if you feel strongly about it.

Reply:
Yea, I'm sure it is more common in those languages although we're not writing those languages. "out-parameters" don't exist in javascript and the popularity of those languages peaked while people still largely wrote imperative code and tried to optimized for computation or memory both of which are lesser concerns now. Javascript, in particular React, is much different, more functional and declarative.

Than being said, I can't say it would be better to change and follow patterns of rest of repo because there are such as mix of functional / non-functional stuff that it's not very clear anymore. I don't know what to feel strongly about anymore. I would say leave it unless you also want to change it

Reply:
Since `collectDialogNodes()` is recursively reentrant from multiple callsites -- and `node` changes on each call -- I think it's cleaner and easier to follow to modify the node in the current context as is done here.

I have, however, modified `getTrainDialogs()` to return the dialogs rather than adding to the collection parameter.